modin/numpy/arr.py

```diff
@@ -0,0 +1,1488 @@
+                f"Modin only supports scalar or modin.numpy.array `where` parameter, not `where` parameter of type {type(where)}"
+            )
+        raise TypeError(
+            f"Cannot cast array data from {where.dtype} to dtype('bool') according to the rule 'safe'"
+        )
+
+
+def check_how_broadcast_to_output(arr_in: "array", arr_out: "array"):
+    if not isinstance(arr_out, array):
+        raise TypeError("return arrays must be of modin.numpy.array type.")
+    if arr_out._ndim == arr_in._ndim and arr_out.shape != arr_in.shape:
+        raise ValueError(
+            f"non-broadcastable output operand with shape {arr_out.shape} doesn't match the broadcast shape {arr_in.shape}"
+        )
+    elif arr_out._ndim == arr_in._ndim:
+        return "broadcastable"
+    elif arr_out._ndim == 1:
+        if prod(arr_in.shape) == arr_out.shape[0]:
+            return "flatten"
+        else:
+            raise ValueError(
+                f"non-broadcastable output operand with shape {arr_out.shape} doesn't match the broadcast shape {arr_in.shape}"
+            )
+    elif arr_in._ndim == 1:
+        if prod(arr_out.shape) == arr_in.shape[0]:
+            return "reshape"
+        else:
+            raise ValueError(
+                f"non-broadcastable output operand with shape {arr_out.shape} doesn't match the broadcast shape {arr_in.shape}"
+            )
+
+
+def fix_dtypes_and_determine_return(
+    query_compiler_in, _ndim, dtype=None, out=None, where=True
+):
+    if dtype is not None:
+        query_compiler_in = query_compiler_in.astype(
+            {col_name: dtype for col_name in query_compiler_in.columns}
+        )
+    result = array(_query_compiler=query_compiler_in, _ndim=_ndim)
+    if out is not None:
+        out = try_convert_from_interoperable_type(out)
+        broadcast_method = check_how_broadcast_to_output(result, out)
+        result._query_compiler = result._query_compiler.astype(
+            {col_name: out.dtype for col_name in result._query_compiler.columns}
+        )
+        if broadcast_method == "flatten":
+            result = result.flatten()
+        elif broadcast_method != "broadcastable":
+            # TODO(RehanSD): Replace this when reshape is implemented.
+            raise NotImplementedError("Reshape is currently not supported in Modin.")
+        if isinstance(where, array):
+            out._query_compiler = where.where(result, out)._query_compiler
+        elif where:
+            out._query_compiler = result._query_compiler
+        return out
+    if isinstance(where, array) and out is None:
+        from array_creation import zeros_like
+
+        out = zeros_like(result).astype(dtype if dtype is not None else result.dtype)
+        out._query_compiler = where.where(result, out)._query_compiler
+        return out
+    elif not where:
+        from .array_creation import zeros_like
+
+        return zeros_like(result)
+    return result
+
+
+def find_common_dtype(dtypes):
```

Comment:
Why not use [numpy.common_type](https://numpy.org/doc/stable/reference/generated/numpy.find_common_type.html)?

Reply:
Fixed!

Reply:
numpy.common_type does not resolve dtypes how we'd like it to. Take the following example:

```python
In [1]: import numpy

In [2]: import pandas

In [3]: df = pandas.DataFrame([[1, '2'], [3, '4']])

In [6]: numpy.find_common_type(df.dtypes.values, [])
Out[6]: dtype('O')

In [7]: numpy.array([[1, '2'], [3, '4']]).dtype
Out[7]: dtype('<U21')
```
the `find_common_dtype` method I wrote uses numpy.promote_types under the hood, and gives us the correct dtype when we our query compiler has mixed types like the df in the example.

Reply:
And we don't use `promote_types` directly since it only accepts two types at a time, so this method I wrote is literally just a wrapper that tree reduces on a list of dtypes.